### DISCUSSION TITLE
How to update status after function returns?

### DISCUSSION THREAD

**Discussion by benfrain** [nvim-lualine/lualine.nvim#765] [Q&A] [answered]
This is perhaps just a lua question. Apologies if that is the case.

I have a function that shows search results:

```lua
function searchResult(quick)
  if vim.v.hlsearch == 0 then
    return ""
  end
  local last_search = vim.fn.getreg("/")
  if not last_search or last_search == "" then
    return ""
  end
  local searchcount = vim.fn.searchcount({ maxcount = 0 })
  return last_search .. "(" .. searchcount.current .. "/" .. searchcount.total .. ")"
end
```

When I jump between search results, the statusline is slow to update the current value. I wondered if this is just because of a delay of the status being redrawn? If so, how can I run a `vim.cmd('redrawstatus')` each time that `searchResult` runs to update instantly?

**Comment by shadmansaleh** [comment, marked as answer]
Your function only gets called when statusline is being refreshed. So calling refresh at end of this doesn't make sense as we are currently in middle of a refresh.

You can reduce how much time lualine waits between refresh with refresh option. By default lualine refreshes every 1 second. You can reduce this time to get more responsive statusline. See [here](https://github.com/nvim-lualine/lualine.nvim/wiki/FAQ#my-statuslinetablinewinbar-updates-infrequently).

In general if you want to refresh lualine on some external event like you pressing n/N you can call `require('lualine').refresh()` but just not inside a lualine component.

**Reply by benfrain** [reply]
Thanks so much for the reply. I understand. Is there any detrimental downside to having a faster update on Lualine? 

Your info on calling .refresh is interesting though as pressing n/N for this use case is infrequent so may make more sense to add it to the key mapping??

**Reply by shadmansaleh** [reply]
> Is there any detrimental downside to having a faster update on Lualine?

The obvious one is the more frequently you refresh the higher cpu utilization you'll have. Lualine still refreshes in sub mili-second so that's not much of a concern on modern machines as long as you set refresh to a reasonable value. For example on my machine with refresh set to 50 ms I get around 2% cpu utilization on one thread while with default 1000ms it sits at pretty much 0%

>Your info on calling .refresh is interesting though though as pressing n/N for this use case is infrequent so may make more sense to add it to the key mapping??

That's what I was hinting at. You can map n/N keys so they trigger search next/prev and then trigger lualine refresh.

**Reply by benfrain** [reply]
Perfect, thanks 🙏

**Comment by benfrain** [comment]
When I set my keymap like this (in Lualine.lua)

```lua
vim.keymap.set("n", "n", function()
  vim.cmd("normal! nzz")
  require("lualine").refresh()
end)
```

I get the following error: attempt to call field 'refresh' (a nil value)

Any idea what I am doing wrong?

**Reply by shadmansaleh** [reply]
Copy pasting this works for me.

Are you on lualine latest master ? That's only reason I can think of to get that error.

Also I'd add `silent` to normal command. `vim.cmd("silent normal! nzz")`
